### PR TITLE
fix:stop the timer when closing the connetion

### PR DESCRIPTION
**Pull request by wubbalubbaaa**
If I have set the idletime , the timer won't be stoped when the connection closed . The timingwheel still holds the reference  to the connection. The connection will not be recyled until it is  expired , which may lead memory leak .
Especially  when the  library  is designed to handle large numbers of connections in websocket , which means mass connections and long idletime

### CODE REVIEW

**Comment by Allenxuxu** [comment]
Thank you very much for finding this bug. 
But there are still some problems with this PR. The use of `conn.timer` is not thread-safe，maybe you just need to check  if connection is closed in `conn.closeTimeoutConn`, or make `conn.timer` to be thread-safe. Both ways have their own advantages.

**Comment by Allenxuxu** [comment]
Thank you very much for finding this bug. 
But there are still some problems with this PR. The use of `conn.timer` is not thread-safe，maybe you just need to check  if connection is closed in `conn.closeTimeoutConn`, or make `conn.timer` to be thread-safe. Both ways have their own advantages.

**Comment by wubbalubbaaa** [comment]
> Thank you very much for finding this bug. But there are still some problems with this PR. The use of `conn.timer` is not thread-safe，maybe you just need to check if connection is closed in `conn.closeTimeoutConn`, or make `conn.timer` to be thread-safe. Both ways have their own advantages.


i have added  a mutex in `Connection` struct to avoid the data race.  i think it may work now

**Comment by Allenxuxu** [comment]
> > Thank you very much for finding this bug. But there are still some problems with this PR. The use of `conn.timer` is not thread-safe，maybe you just need to check if connection is closed in `conn.closeTimeoutConn`, or make `conn.timer` to be thread-safe. Both ways have their own advantages.
> 
> i have added a mutex in `Connection` struct to avoid the data race. i think it may work now

I think atomic.Value is better🤔

**Comment by wubbalubbaaa** [comment]
> > > Thank you very much for finding this bug. But there are still some problems with this PR. The use of `conn.timer` is not thread-safe，maybe you just need to check if connection is closed in `conn.closeTimeoutConn`, or make `conn.timer` to be thread-safe. Both ways have their own advantages.
> > 
> > 
> > i have added a mutex in `Connection` struct to avoid the data race. i think it may work now
> 
> I think atomic.Value is better🤔

yes, you are right , I'm not experienced enough😂

File: connection.go [hunk 1]
```diff
@@ -282,14 +287,18 @@ func (c *Connection) handleClose(fd int) {
...
-
 		c.callBack.OnClose(c)
 		if err := unix.Close(fd); err != nil {
 			log.Error("[close fd]", err)
 		}
-
 		ringbuffer.PutInPool(c.inBuffer)
 		ringbuffer.PutInPool(c.outBuffer)
+		if c.idleTime > 0 {
+			timer, ok := c.timer.Load().(*timingwheel.Timer)
```
Comment (Allenxuxu) [inline connection.go:297]:
```
if v := c.timer.Load(); v != nil {
  timer, ok := v.(*timingwheel.Timer)
```

File: connection.go [hunk 2]
```diff
@@ -282,14 +287,18 @@ func (c *Connection) handleClose(fd int) {
...
 		c.loop.DeleteFdInLoop(fd)
-
 		c.callBack.OnClose(c)
 		if err := unix.Close(fd); err != nil {
 			log.Error("[close fd]", err)
 		}
-
 		ringbuffer.PutInPool(c.inBuffer)
 		ringbuffer.PutInPool(c.outBuffer)
+		if c.idleTime > 0 {
```
Comment (Allenxuxu) [inline connection.go:296]: Maybe we don't need it now, because we just need to check `c.timer`

**Review by Allenxuxu — CHANGES_REQUESTED** [review]

File: connection.go (hunk 2 above)
Comment (wubbalubbaaa) [reply connection.go]: done, have simplified the code

**Review by Allenxuxu — APPROVED** [review]

**Comment by Allenxuxu** [comment]
Thank you very much. I merged it.

**Pull request merged**